### PR TITLE
Restructure manager and views

### LINKED ISSUE #22: Unified structure for course endpoints

### LINKED ISSUE #36: the download button downloads a compressed file with (report.csv, crash_datails)

### LINKED ISSUE #61: Change manager/views to work with the new graders

### PR DESCRIPTION
**Pull request by amohamed97**
* Manager now acts as a proxy to multiple graders, each grader has the all the previous manager's functionality but customized to a certain course type
* Views endpoints changed to cope with the manager's changes
* Downloadable result is now a compressed file

resolves #61 , resolves #36 , resolves #22

### CODE REVIEW

File: graderx/graders/manager.py [hunk 1]
```diff
@@ -1,93 +1,86 @@
...
-
-def save_single_submission(lab_id, submission_file, file_name):
-    curr_dir = str(Path(__file__).parent.resolve())
-    Path(f'{curr_dir}/courses/cc451/app/{lab_id}/submissions/2020/{file_name}').write_bytes(
-        submission_file.getbuffer())
-
-
-class ArchiveDamagedError(Exception):
+
+def compressed_results(course_name, lab):
```
Comment (Hagar-Usama) [inline graderx/graders/manager.py:73]: why do we need to compress submissions?

File: graderx/graders/stdout_graders/c/c_grader.py [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+    return Path(__file__).joinpath(
+        f'../../../courses/{course}/labs/{lab}').resolve()
+
+
+def run_grader(course, lab):
+    """
+    Compiles all the submissions in [lab_path/submissions/] directory, runs compiled submissions
+    with stdin of the lab's test cases, then builds a dictionary that has all the submissions 
+    with the submissions id as the key and the results of this submissions as the value 
+    example dictionray:
```
Comment (Hagar-Usama) [inline graderx/graders/stdout_graders/c/c_grader.py:23]: New Example:
```json
{
        "id": "1234_5678",
        "passed": ["1", "2"],
        "failed": [
            {
                "tc_id": "3",
                "output": "STRING THAT IS STUDENT OUTPUT AFTER COMPILATION FOR THE GIVEN INPUT",
                "expected": "STRING THAT IS THE EXPECTED OUTPUT FOR THE GIVEN INPUT"
            }
        ]
    },
    {
        "id": "9876_5432",
        "passed": ["1"],
        "failed": [
            {
                "tc_id": "2",
                "output": "STRING THAT IS STUDENT OUTPUT AFTER COMPILATION FOR THE GIVEN INPUT",
                "expected": "STRING THAT IS THE EXPECTED OUTPUT FOR THE GIVEN INPUT"
            },
            {
                "tc_id": "3",
                "output": "STRING THAT IS STUDENT OUTPUT AFTER COMPILATION FOR THE GIVEN INPUT",
                "expected": "STRING THAT IS THE EXPECTED OUTPUT FOR THE GIVEN INPUT"
            }
        ]
    }
}
```

File: graderx/graders/unittest_graders/pytest/lib/submissions_extraction.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+            f.unlink()
+
+
+def extract_file(file_path, verbosity=-1):
+    patoolib.extract_archive(
+        file_path, outdir=file_path.parent, verbosity=verbosity)
+    os.remove(file_path)
+
+
+def extract_submissions(dest_directory, submissions_file,  verbosity=0):
```
Comment (Hagar-Usama) [inline graderx/graders/unittest_graders/pytest/lib/submissions_extraction.py:21]: why do we have 2 files or sumbissions_extraction?
`graderx/graders/stdout_graders/c/lib/submissions_extraction.py`
`graderx/graders/unittest_graders/pytest/lib/submissions_extraction.py`

File: graderx/graders/unittest_graders/pytest/pytest_grader.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+def run_grader(course, lab):
+    """
+    Runs the tests in the given lab's corresponding pytest test file 
+    which is test_run_grader.py in each lab directory  
+    then saves the output in a file
+    then pipes the pytest output file to console_log_parser 
+    which will parse the pytest output to create 4 files
+    1- [TIMESTAMP]_[LABNAME]_crash_details.txt which contains 
+    detailed pytest output of only the failed tests for all the submissions
+    2- [TIMESTAMP]_[LABNAME]_grade_summary.csv which contains
```
Comment (Hagar-Usama) [inline graderx/graders/unittest_graders/pytest/pytest_grader.py:26]: is grade_summary.csv related (somehow) to [lab]_result_summary.txt (compute_results)?
shall compute_results return [lab]_result_summary.csv instead?

**Review by Hagar-Usama — APPROVED** [review]

**Comment by Hagar-Usama** [comment]
{
    "url": "url",
    "stauts": "status",
    "command": "command"
}

**Review by amrable — APPROVED** [review]

**Review by amroakmal — APPROVED** [review]

**Pull request merged**